File: preflight/layer2/typescript/_shared.js
'use strict';
/**
 * Shared helpers for the TypeScript/Node.js Layer 2 probes (probe.js +
 * probe_sdk.js). Mirrors preflight/layer2/python/_shared.py's structure and
 * division of labor -- centralized so probes never carry two independently-
 * drifting copies of the same fragment-emission / classification /
 * redaction logic (one shared vocabulary across every language).
 *
 * Plain CommonJS, zero dependencies, zero build step -- `require('./_shared')`
 * works with nothing beyond the `node` binary itself, matching every other
 * tier-1 probe's "no build step" ethos in this project. (Tier 2's SDK usage
 * needs its own npm install, same precedent as Java's tier-2 needing its own
 * `javac`/Maven step -- see probe_sdk.js.)
 */

const fs = require('fs');
const net = require('net');
const tls = require('tls');
const { URL } = require('url');

const RUNTIME = 'typescript';
const CONNECT_TIMEOUT_MS = 10000;

// Matches inline URL credentials (scheme://user:pass@host) so they can be
// masked out of any text before it's emitted (credentials embedded in a URL
// must never reach output; mirrored from _shared.py's scrub_url_creds).
const URL_CREDS_RE = /(\w+:\/\/)[^/@\s:]+:[^/@\s]+@/g;

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

function scrubUrlCreds(text) {
  if (!text) return text;
  return String(text).replace(URL_CREDS_RE, '$1****:****@');
}

function eprint(...args) {
  // eslint-disable-next-line no-console
  console.error(...args);
}

function fragment(target, verdict, errorClass, detail, storeLabel) {
  // Scrub URL credentials from the detail as a universal backstop -- every
  // emitted fragment funnels through here, so a proxy password embedded in an
  // exception string can't leak to stdout regardless of which code path
  // built the detail (mirrors probe.py's fragment()).
  return {
    runtime: RUNTIME,
    trustStoreExercised: storeLabel || '',
    target: target || '',
    verdict,
    errorClass,
    detail: scrubUrlCreds(detail),
  };
}

function crashFragment(detail) {
  return fragment('', 'probe-error', 'PROBE_CRASHED', 'probe crashed: ' + detail);
}

/**
 * Whether to surface extra diagnostic fragments hidden by default (e.g. the
 * Console-URL normalization notice) -- useful to the operator/trainer, but
 * confusing noise for participants. Set by the launcher via
 * CAMUNDA_PREFLIGHT_VERBOSE (from the Go binary's --verbose), or by passing
 * --verbose when running a probe standalone.
 */
function isVerbose() {
  if (process.argv.slice(2).includes('--verbose')) return true;
  const v = (process.env.CAMUNDA_PREFLIGHT_VERBOSE || '').trim().toLowerCase();
  return v === '1' || v === 'true' || v === 'yes';
}

/**
 * Opt-in switch for probe_sdk.js's proxy-tunneling `fetch` override (see
 * tunneledFetch in probe_sdk.js). OFF by default -- the real SDK's own
 * `fetch` has zero proxy handling and silently connects direct, which is
 * exactly what this tool's default behavior mirrors on purpose (matching
 * what a cohort's own unmodified job-worker code would do). Set
 * CAMUNDA_TS_PROXY_SUPPORT=1 (forwarded from the Go binary's
 * --ts-proxy-support) or pass --ts-proxy-support standalone to opt into the
 * tunneled check instead -- see RUNBOOK's TypeScript section for what that
 * does and does not prove.
 */
function tsProxySupportEnabled() {
  if (process.argv.slice(2).includes('--ts-proxy-support')) return true;
  const v = (process.env.CAMUNDA_TS_PROXY_SUPPORT || '').trim().toLowerCase();
  return v === '1' || v === 'true' || v === 'yes';
}

/**
 * Mirrors the Go binary's network-vs-full decision so probes never diverge
 * from it (network mode must be credential-free -- status only, no
 * authenticated topology, no token). An explicit
 * mode (passed by the launcher via CAMUNDA_PREFLIGHT_MODE) wins; when unset
 * (probe run standalone by hand) fall back to creds-presence auto-detect,
 * the same default the Go binary uses.
 */
function resolveIsFull(mode, hasCreds) {
  const m = (mode || '').trim().toLowerCase();
  if (m === 'network') return false;
  if (m === 'full') return true;
  return hasCreds;
}

function getProxyUrl() {
  for (const name of ['HTTPS_PROXY', 'https_proxy', 'HTTP_PROXY', 'http_proxy']) {
    const v = process.env[name];
    if (v) return v;
  }
  return null;
}

function maskProxy(proxyUrl) {
  if (!proxyUrl) return '';
  let u;
  try {
    u = new URL(proxyUrl);
  } catch (e) {
    return proxyUrl;
  }
  if (u.username || u.password) {
    u.username = '****';
    u.password = '****';
    return u.toString();
  }
  return proxyUrl;
}

/**
 * Same config-source precedence as the Go binary: an explicit
 * CAMUNDA_REST_ADDRESS host wins over CAMUNDA_REGION.
 */
function resolveApiHost() {
  const restAddress = (process.env.CAMUNDA_REST_ADDRESS || '').trim();
  const region = (process.env.CAMUNDA_REGION || '').trim() || 'bru-2';

  if (restAddress) {
    const candidate = restAddress.includes('://') ? restAddress : 'https://' + restAddress;
    try {
      const u = new URL(candidate);
      if (u.hostname) return u.hostname;
    } catch (e) {
      // fall through to the region-derived default
    }
  }
  return region + '.api.camunda.io';
}

/**
 * Rebuild a canonical https://<host>/<clusterId> REST base, mirroring the Go
 * binary's hostset.parseExplicitHost tolerance (UUID-anywhere-in-path +
 * authority-port stripping) and Python's _shared.normalize_rest_base /
 * Java's Shared.normalizeRestBase.
 *
 * This exists because the real Camunda TypeScript SDK does NOT tolerate the
 * exact string Camunda Console tells users to copy either -- verified live
 * against the SDK's own config-hydration source (dist/chunk-WSCXETVI.js): it
 * only appends '/v2' when the path doesn't already end in it, with no UUID
 * extraction or stray-segment stripping.
 * Console's copy-paste form embeds a stray ':443' path segment
 * (https://<host>/:443/<clusterId>/v2/); passed straight through, the SDK
 * would hit /:443/<clusterId>/v2/status -> Cloudflare 'default backend - 404'
 * -- the SAME known gap already confirmed for the Python and Java SDKs.
 *
 * Returns { restBase, host, wasNormalized }. wasNormalized is true when the
 * raw input carried stray path segments (the exact case a hand-configuring
 * participant would trip over), so the caller can WARN about it.
 */
function normalizeRestBase() {
  const raw = (process.env.CAMUNDA_REST_ADDRESS || '').trim();
  const region = (process.env.CAMUNDA_REGION || '').trim() || 'bru-2';
  const clusterIdEnv = (process.env.CAMUNDA_CLUSTER_ID || '').trim();

  if (!raw) {
    const host = region + '.api.camunda.io';
    if (clusterIdEnv) {
      return { restBase: 'https://' + host + '/' + clusterIdEnv, host, wasNormalized: false };
    }
    return { restBase: 'https://' + host, host, wasNormalized: false };
  }

  const candidate = raw.includes('://') ? raw : 'https://' + raw;
  let parsed;
  try {
    parsed = new URL(candidate);
  } catch (e) {
    return { restBase: candidate, host: region + '.api.camunda.io', wasNormalized: false };
  }
  const host = parsed.hostname || region + '.api.camunda.io';

  const segments = parsed.pathname.split('/').filter(Boolean);
  let clusterId = segments.find((s) => UUID_RE.test(s));
  if (!clusterId) clusterId = clusterIdEnv;

  if (!clusterId) {
    // No UUID anywhere -- return the input untouched rather than guessing;
    // the SDK will surface its own error and the native probe still runs.
    return { restBase: candidate, host, wasNormalized: false };
  }

  const canonical = 'https://' + host + '/' + clusterId;
  // Non-canonical if the raw path carried anything beyond the clusterId and
  // an optional trailing "v2" (e.g. a stray ":443"/"443" Console segment), or
  // if the authority carried an explicit port.
  const straySegments = segments.filter((s) => s !== clusterId && s !== 'v2');
  const wasNormalized = straySegments.length > 0 || parsed.port !== '';
  return { restBase: canonical, host, wasNormalized };
}

// error.code values mirror the Go binary's classifyDialError (httpclient.go)
// and probe.py/Shared.java's classification -- same DNS/refused/timeout/cert
// distinctions, expressed via Node's actual documented error.code values
// (verified live: net.createConnection against a non-existent hostname threw
// code ENOTFOUND; against a closed port threw
// ECONNREFUSED; a certificate-trust failure threw
// UNABLE_TO_GET_ISSUER_CERT_LOCALLY). ETIMEDOUT/ABORT_ERR are set by this
// project's own timeout wrappers (Node's raw net/tls sockets don't
// self-assign a `.code` on a bare 'timeout' event; AbortSignal.timeout()
// used by probe_sdk.js does set `.code = 'ABORT_ERR'`).
const DNS_CODES = new Set(['ENOTFOUND', 'EAI_AGAIN', 'ENODATA']);
const REFUSED_CODES = new Set(['ECONNREFUSED', 'ECONNABORTED', 'EACCES', 'EHOSTUNREACH', 'ENETUNREACH']);
const TIMEOUT_CODES = new Set(['ETIMEDOUT', 'ABORT_ERR', 'UND_ERR_CONNECT_TIMEOUT']);
const CERT_CODES = new Set([
  'UNABLE_TO_VERIFY_LEAF_SIGNATURE',
  'UNABLE_TO_GET_ISSUER_CERT_LOCALLY',
  'UNABLE_TO_GET_ISSUER_CERT',
  'CERT_HAS_EXPIRED',
  'CERT_NOT_YET_VALID',
  'DEPTH_ZERO_SELF_SIGNED_CERT',
  'SELF_SIGNED_CERT_IN_CHAIN',
  'ERR_TLS_CERT_ALTNAME_INVALID',
  'CERT_UNTRUSTED',
  'ERR_OSSL_X509_KEY_VALUES_MISMATCH',
]);

/**
 * Walks err.cause (Node's fetch/undici TypeError chaining -- e.g. the SDK's
 * NetworkSdkError sets `.cause` to the original error) the same way
 * _shared.py walks __cause__/__context__ and Shared.java walks getCause().
 * Also descends into AggregateError.errors[0] (Node's Happy-Eyeballs
 * dual-stack connect can throw an AggregateError wrapping multiple per-
 * address failures), so classification still works for that shape.
 */
function errorChain(err) {
  const chain = [];
  const seen = new Set();
  let cur = err;
  while (cur && typeof cur === 'object' && !seen.has(cur)) {
    seen.add(cur);
    chain.push(cur);
    if (Array.isArray(cur.errors) && cur.errors.length) {
      cur = cur.errors[0];
      continue;
    }
    cur = cur.cause;
  }
  if (chain.length === 0) chain.push(err);
  return chain;
}

/**
 * Classify a connection/TLS error into [errorClass, detail]. Structural
 * (error.code) checks first, then substring matching on the chained error
 * text for wrapped/opaque cases -- the same layered approach used across
 * every language in this project.
 */
function classifyTransportError(err) {
  const chain = errorChain(err);

  for (const e of chain) {
    const code = e && e.code;
    if (!code) continue;
    if (DNS_CODES.has(code)) {
      return ['DNS_FAIL', 'hostname did not resolve: ' + (e.message || e)];
    }
    if (REFUSED_CODES.has(code)) {
      return ['CONNECT_REFUSED', 'connection refused -- port 443 likely blocked by firewall: ' + (e.message || e)];
    }
    if (TIMEOUT_CODES.has(code)) {
      return ['CONNECT_TIMEOUT', 'connection timed out -- port 443 likely blocked/dropped by firewall: ' + (e.message || e)];
    }
    if (CERT_CODES.has(code)) {
      return ['TLS_HANDSHAKE_FAIL', 'certificate not trusted: ' + (e.message || e)];
    }
  }

  const text = chain
    .map((e) => (e && (e.message || e.toString())) || '')
    .join(' | ')
    .toLowerCase();
  if (/timed out|timeout/.test(text)) {
    return ['CONNECT_TIMEOUT', 'connection timed out -- port 443 likely blocked/dropped by firewall: ' + chain[0]];
  }
  if (/refused|forbidden by its access permissions/.test(text)) {
    return ['CONNECT_REFUSED', 'connection refused -- port 443 likely blocked by firewall: ' + chain[0]];
  }
  if (/getaddrinfo|nodename nor servname|name or service not known|dns/.test(text)) {
    return ['DNS_FAIL', 'hostname did not resolve: ' + chain[0]];
  }
  if (/cert|ssl|tls|handshake/.test(text)) {
    return ['TLS_HANDSHAKE_FAIL', 'TLS handshake failed: ' + chain[0]];
  }
  return ['CONNECT_REFUSED', 'connection failed: ' + chain[0]];
}

/**
 * Raw plain-TCP connect with a manual timeout (net.Socket has no built-in
 * connect timeout for a 'connect' event that never fires). Shared by both
 * probe.js's own tier-1 checks AND probe_sdk.js's proxy-aware fetch override
 * (see connectViaProxy/tlsHandshake below) -- moved here rather than kept
 * private to probe.js so the two probes never carry two independently-
 * drifting copies of the exact same socket/tunnel/trust logic.
 */
function connectPlain(host, port, timeoutMs) {
  return new Promise((resolve, reject) => {
    const sock = net.createConnection({ host, port });
    let settled = false;
    const timer = setTimeout(() => {
      if (settled) return;
      settled = true;
      sock.destroy();
      const e = new Error('connection timed out after ' + timeoutMs + 'ms');
      e.code = 'ETIMEDOUT';
      reject(e);
    }, timeoutMs);
    sock.once('connect', () => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve(sock);
    });
    sock.once('error', (e) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      reject(e);
    });
  });
}

class ProxyError extends Error {
  constructor(statusLine) {
    super(statusLine || 'no response from proxy');
    this.statusLine = statusLine || '';
  }
}

/**
 * Opens an HTTP CONNECT tunnel through proxyUrl to host:port -- raw socket,
 * manual CONNECT handshake, Basic proxy-auth from the URL's userinfo,
 * upgrading to TLS only after the tunnel is established. Confirmed live
 * (probe.js's original research): Node's tls.connect()/https/global fetch do
 * NOT auto-tunnel through HTTPS_PROXY/HTTP_PROXY, and the real SDK itself has
 * ZERO proxy-env-var handling of its own -- this manual tunnel is the only
 * way this project's proxy support works for Node at all, for either tier
 * (probe_sdk.js's proxy-aware fetch override reuses this exact function so
 * tier 2 stops silently bypassing --proxy the way global fetch does).
 */
function connectViaProxy(proxyUrl, host, port, timeoutMs) {
  return new Promise((resolve, reject) => {
    let proxy;
    try {
      proxy = new URL(proxyUrl);
    } catch (e) {
      reject(e);
      return;
    }
    const proxyPort = proxy.port ? Number(proxy.port) : proxy.protocol === 'https:' ? 443 : 80;

    connectPlain(proxy.hostname, proxyPort, timeoutMs)
      .then((rawSock) => {
        const sock = proxy.protocol === 'https:' ? tls.connect({ socket: rawSock, servername: proxy.hostname }) : rawSock;

        const lines = ['CONNECT ' + host + ':' + port + ' HTTP/1.1', 'Host: ' + host + ':' + port];
        if (proxy.username) {
          const user = decodeURIComponent(proxy.username);
          const pass = decodeURIComponent(proxy.password || '');
          const cred = Buffer.from(user + ':' + pass).toString('base64');
          lines.push('Proxy-Authorization: Basic ' + cred);
        }

        let buf = Buffer.alloc(0);
        let settled = false;

        const timer = setTimeout(() => {
          if (settled) return;
          settled = true;
          sock.destroy();
          const e = new Error('proxy CONNECT tunnel timed out after ' + timeoutMs + 'ms');
          e.code = 'ETIMEDOUT';
          reject(e);
        }, timeoutMs);

        const onData = (chunk) => {
          buf = Buffer.concat([buf, chunk]);
          const idx = buf.indexOf('\r\n\r\n');
          if (idx === -1 || settled) return;
          settled = true;
          clearTimeout(timer);
          sock.removeListener('data', onData);
          sock.removeListener('error', onError);
          const statusLine = buf.slice(0, buf.indexOf('\r\n')).toString('latin1').trim();
          if (!/ 200(\s|$)/.test(statusLine)) {
            sock.destroy();
            reject(new ProxyError(statusLine));
            return;
          }
          resolve(sock);
        };
        const onError = (e) => {
          if (settled) return;
          settled = true;
          clearTimeout(timer);
          reject(e);
        };
        sock.on('data', onData);
        sock.on('error', onError);
        sock.write(lines.join('\r\n') + '\r\n\r\n');
      })
      .catch(reject);
  });
}

function tlsHandshake(rawSocket, host, trust, timeoutMs) {
  return new Promise((resolve, reject) => {
    const tlsSock = tls.connect({
      socket: rawSocket,
      servername: host,
      ca: trust.ca,
      rejectUnauthorized: true,
    });
    const timer = setTimeout(() => {
      tlsSock.destroy();
      const e = new Error('TLS handshake timed out after ' + timeoutMs + 'ms');
      e.code = 'ETIMEDOUT';
      reject(e);
    }, timeoutMs);
    tlsSock.once('secureConnect', () => {
      clearTimeout(timer);
      resolve(tlsSock);
    });
    tlsSock.once('error', (e) => {
      clearTimeout(timer);
      reject(e);
    });
  });
}

/**
 * Bundles the trust material (`ca: undefined` means Node's own default
 * store) with a human-readable label, plus an optional pre-formed WARN
 * fragment (e.g. an unreadable custom-CA file). Shared by probe.js's tier-1
 * checks and probe_sdk.js's proxy-aware fetch override -- both tiers must
 * make the IDENTICAL trust decision for the same CAMUNDA_MTLS_CA_PATH value,
 * or Layer 2 re-introduces the exact false-green this whole probe exists to
 * prevent. REPLACE, not append -- verified against the real SDK's own
 * published source (see probe.js's original module doc comment for the full
 * verified reasoning, including a live empirical test): once a custom CA is
 * set, ONLY that CA is trusted, not "OS store + custom CA".
 */
function buildTrustContext() {
  const customCaPath = (process.env.CAMUNDA_MTLS_CA_PATH || '').trim();
  if (!customCaPath) {
    return {
      ca: undefined,
      label: 'the default certificate store',
      warn: null,
    };
  }

  let pem;
  try {
    pem = fs.readFileSync(customCaPath, 'utf8');
  } catch (e) {
    return {
      ca: undefined,
      label: 'the default certificate store (your custom certificate path is unreadable -- NOT applied)',
      warn: fragment(
        resolveApiHost() + ' (config)',
        'WARN',
        'CONFIG_ERROR',
        'CAMUNDA_MTLS_CA_PATH is set to ' + customCaPath + ', but that file could not be read (' + e.message + ') -- proceeding WITHOUT the custom CA.'
      ),
    };
  }

  return {
    ca: pem,
    label: 'your custom certificate (' + customCaPath + ')',
    warn: null,
  };
}

module.exports = {
  RUNTIME,
  CONNECT_TIMEOUT_MS,
  scrubUrlCreds,
  eprint,
  fragment,
  crashFragment,
  isVerbose,
  tsProxySupportEnabled,
  resolveIsFull,
  getProxyUrl,
  maskProxy,
  resolveApiHost,
  normalizeRestBase,
  classifyTransportError,
  errorChain,
  connectPlain,
  ProxyError,
  connectViaProxy,
  tlsHandshake,
  buildTrustContext,
};

File: preflight/layer2/python/_shared.py
"""Shared helpers for the Python Layer 2 probes (probe.py + probe_sdk.py).

Centralized so the two probes never carry two independently-drifting copies
of the same transport-error classification — the exact kind of divergence
this whole project exists to prevent by keeping one shared error-classification
vocabulary across every probe language. Both the raw-socket probe and the
SDK-based probe funnel their connection errors through
classify_transport_error() below.
"""
import os
import re
import sys
from urllib.parse import urlparse

RUNTIME = "python"

_UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.IGNORECASE)

# Matches inline URL credentials (scheme://user:pass@host) so they can be
# masked out of any text before it's emitted.
_URL_CREDS_RE = re.compile(r"(\w+://)[^/@\s:]+:[^/@\s]+@")


def scrub_url_creds(text):
    """Mask user:pass@ credentials in any URL embedded in text. Applied to
    every fragment detail before emission, so a proxy password carried in an
    httpx/socket exception string can never reach stdout in the clear."""
    if not text:
        return text
    return _URL_CREDS_RE.sub(r"\1****:****@", text)


def silence_stdlib_logging():
    """Defense-in-depth: suppress ALL Python stdlib logging in the probe
    process. The SDK's own NullLogger stops its own logger
    abstraction, but a transitive dependency (httpx/httpcore) — or a future SDK
    version — could log an unmasked client_id via the stdlib `logging` module,
    which NullLogger doesn't intercept. The probe reports through its own JSON
    (stdout) and eprint (stderr) channels and does not rely on stdlib logging,
    so nothing of value is lost by disabling it outright."""
    import logging

    logging.disable(logging.CRITICAL)


def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)


def fragment(target, verdict, error_class, detail, store_label=""):
    # Scrub URL credentials from the detail as a universal backstop — every
    # emitted fragment funnels through here, so a proxy password embedded in an
    # exception string can't leak to stdout regardless of which code path
    # built the detail.
    return {
        "runtime": RUNTIME,
        "trustStoreExercised": store_label,
        "target": target,
        "verdict": verdict,
        "errorClass": error_class,
        "detail": scrub_url_creds(detail),
    }


def crash_fragment(detail):
    return fragment("", "probe-error", "PROBE_CRASHED", "probe crashed: %s" % detail)


def is_verbose():
    """Whether to surface extra diagnostic fragments hidden by default (e.g. the
    Console-URL normalization notice) — useful to the operator/trainer, but
    confusing noise for participants. Set by the launcher via
    CAMUNDA_PREFLIGHT_VERBOSE (from the Go binary's --verbose), or by passing
    --verbose when running the probe standalone."""
    if "--verbose" in sys.argv[1:]:
        return True
    return os.environ.get("CAMUNDA_PREFLIGHT_VERBOSE", "").strip().lower() in ("1", "true", "yes")


def resolve_is_full(mode, has_creds):
    """Mirror the Go binary's network-vs-full decision so probes never diverge
    from it: network mode must be credential-free — status only, no
    authenticated topology, no token. An explicit
    mode (passed by the launcher via CAMUNDA_PREFLIGHT_MODE) wins; when unset
    (probe run standalone by hand) fall back to creds-presence auto-detect,
    the same default the Go binary uses."""
    m = (mode or "").strip().lower()
    if m == "network":
        return False
    if m == "full":
        return True
    return has_creds


def get_proxy_url():
    for name in ("HTTPS_PROXY", "https_proxy", "HTTP_PROXY", "http_proxy"):
        v = os.environ.get(name)
        if v:
            return v
    return None


def mask_proxy(proxy_url):
    if not proxy_url:
        return ""
    p = urlparse(proxy_url)
    if p.username:
        netloc = "****:****@%s" % p.hostname
        if p.port:
            netloc += ":%d" % p.port
        return p._replace(netloc=netloc).geturl()
    return proxy_url


def resolve_api_host():
    """Same config-source precedence as the Go binary:
    an explicit CAMUNDA_REST_ADDRESS host wins over CAMUNDA_REGION."""
    rest_address = os.environ.get("CAMUNDA_REST_ADDRESS", "").strip()
    region = os.environ.get("CAMUNDA_REGION", "").strip() or "bru-2"

    if rest_address:
        candidate = rest_address if "://" in rest_address else "https://" + rest_address
        hostname = urlparse(candidate).hostname
        if hostname:
            return hostname
    return "%s.api.camunda.io" % region


def cluster_edge_detail(status_code):
    """Detail text for CLUSTER_EDGE_404 -- the catch-all for a status the SDK
    reached the real cluster edge with but that isn't specifically handled
    (503 and 401 get their own classifications). Same wording as the Go
    binary's status.go and the Java/C# SDK probes, so the message reads
    identically regardless of which check produced it."""
    return (
        "reached the cluster edge (HTTP %d) but got no valid cluster route -- this typically means our shared "
        "preflight cluster is paused or in a transient edge blip, NOT a problem with your network. Re-run in "
        "~5 minutes; if it persists, contact the training team." % status_code
    )


def has_cluster_target():
    """Whether enough config exists to build a REST base with an actual
    cluster route -- CAMUNDA_REST_ADDRESS carries a UUID somewhere in its
    path, or CAMUNDA_CLUSTER_ID is one itself.

    Mirrors the Go binary's own hostset.Resolve validation. That check runs
    before the binary ever invokes a probe, so it normally catches a missing
    cluster id before any Layer 2 code executes at all -- but this file is
    also runnable standalone, which bypasses it. Without this, normalize_rest_base()
    below silently falls back to a bare https://<region>.api.camunda.io with no
    cluster path, and the SDK's resulting 404 reads as a transient cluster
    problem instead of a missing --cluster-id/--host.
    """
    raw = os.environ.get("CAMUNDA_REST_ADDRESS", "").strip()
    if raw:
        for seg in raw.split("/"):
            if _UUID_RE.match(seg):
                return True
    return bool(_UUID_RE.match(os.environ.get("CAMUNDA_CLUSTER_ID", "").strip()))


def normalize_rest_base():
    """Rebuild a canonical https://<host>/<clusterId> REST base, mirroring the
    Go binary's hostset.parseExplicitHost tolerance (UUID-anywhere-in-path +
    authority-port stripping).

    This exists because the Camunda Python SDK does NOT tolerate the exact
    string Camunda Console tells users to copy: Console's copy-paste form
    embeds a stray ':443' path segment
    (https://<host>/:443/<clusterId>/v2/). The Go binary strips it; the SDK
    appends '/v2' and passes the whole path to httpx, hitting
    /:443/<clusterId>/v2/status -> Cloudflare 'default backend - 404'.
    Passing this normalized value to the SDK via
    its `configuration` param makes the SDK probe test the same canonical
    endpoint every other check does, instead of a false-red 404.

    Returns (rest_base, host, was_normalized). was_normalized is True when the
    raw input carried stray path segments (the exact case a hand-configuring
    participant would trip over), so the caller can WARN about it.
    """
    raw = os.environ.get("CAMUNDA_REST_ADDRESS", "").strip()
    region = os.environ.get("CAMUNDA_REGION", "").strip() or "bru-2"
    cluster_id_env = os.environ.get("CAMUNDA_CLUSTER_ID", "").strip()

    if not raw:
        host = "%s.api.camunda.io" % region
        if cluster_id_env:
            return "https://%s/%s" % (host, cluster_id_env), host, False
        return "https://%s" % host, host, False

    candidate = raw if "://" in raw else "https://" + raw
    parsed = urlparse(candidate)
    host = parsed.hostname or ("%s.api.camunda.io" % region)

    cluster_id = None
    for seg in parsed.path.split("/"):
        if _UUID_RE.match(seg):
            cluster_id = seg
            break
    if not cluster_id:
        cluster_id = cluster_id_env

    if not cluster_id:
        # No UUID anywhere — return the input untouched rather than guessing;
        # the SDK will surface its own error and the native probe still runs.
        return candidate, host, False

    canonical = "https://%s/%s" % (host, cluster_id)
    # Non-canonical if the raw path carried anything beyond the clusterId and
    # an optional trailing "v2" (e.g. a stray ":443"/"443" Console segment),
    # or if the authority carried an explicit port.
    stray_segments = [s for s in parsed.path.split("/") if s and s != cluster_id and s != "v2"]
    was_normalized = bool(stray_segments) or (parsed.port is not None)
    return canonical, host, was_normalized


# Substring rules mirror the Go binary's classifyDialError (httpclient.go) —
# same three confirmed OS wordings for a blocked connection, same DNS/timeout/
# cert-trust distinctions. Kept as text matching (not exception-type
# matching) so this ALSO works for httpx-wrapped errors in probe_sdk.py,
# where the original OS-level exception may be buried in __cause__/__context__
# rather than raised directly.
_DNS_MARKERS = (
    "nodename nor servname", "name or service not known", "getaddrinfo failed",
    "temporary failure in name resolution", "[errno 11001]", "[errno -2]",
    "[errno -3]",
)
_REFUSED_MARKERS = ("connection refused", "actively refused", "forbidden by its access permissions")
_TIMEOUT_MARKERS = ("timed out", "timeout")
_CERT_MARKERS = ("certificate_verify_failed", "certificate verify failed", "unable to get local issuer certificate")


def classify_transport_error(exc):
    """Classify a connection/TLS exception into (errorClass, detail).

    Walks __cause__/__context__ so an httpx-wrapped exception (whose real
    OS-level error is chained, not the top-level type) still classifies the
    same way a raw socket/ssl exception would.
    """
    chain = []
    seen = set()
    cur = exc
    while cur is not None and id(cur) not in seen:
        seen.add(id(cur))
        chain.append(cur)
        cur = cur.__cause__ or cur.__context__

    for e in chain:
        if isinstance(e, __import__("socket").gaierror):
            return "DNS_FAIL", "hostname did not resolve: %s" % e
        if isinstance(e, ConnectionRefusedError):
            return "CONNECT_REFUSED", "connection refused — port 443 likely blocked by firewall: %s" % e
        if isinstance(e, (__import__("socket").timeout, TimeoutError)):
            return "CONNECT_TIMEOUT", "connection timed out — port 443 likely blocked/dropped by firewall: %s" % e

    text = " | ".join(str(e) for e in chain).lower()
    if any(m in text for m in _CERT_MARKERS):
        return "TLS_HANDSHAKE_FAIL", "certificate not trusted: %s" % chain[0]
    if any(m in text for m in _DNS_MARKERS):
        return "DNS_FAIL", "hostname did not resolve: %s" % chain[0]
    if any(m in text for m in _REFUSED_MARKERS):
        return "CONNECT_REFUSED", "connection refused — port 443 likely blocked by firewall: %s" % chain[0]
    if any(m in text for m in _TIMEOUT_MARKERS):
        return "CONNECT_TIMEOUT", "connection timed out — port 443 likely blocked/dropped by firewall: %s" % chain[0]
    if "ssl" in text or "tls" in text or "handshake" in text:
        return "TLS_HANDSHAKE_FAIL", "TLS handshake failed: %s" % chain[0]
    return "CONNECT_REFUSED", "connection failed: %s" % chain[0]
